File: backend/Bruno.Domain/ValueObjects/RegistrationNumber.cs
using System.Text.RegularExpressions;
using Bruno.Domain.Common;

namespace Bruno.Domain.ValueObjects;

/// <summary>
/// South African vehicle registration helpers: normalize and validate plate formats
/// (legacy provincial, newer provincial with suffix, and personalised).
/// </summary>
public static class RegistrationNumber
{
    public const int MinLength = 3;
    public const int MaxLength = 12;

    public const string InvalidMessage = "Registration number must be a valid South African plate.";

    private static readonly Regex LegacyPattern = new(
        @"^[A-Z]{1,3}\d{1,6}$",
        RegexOptions.Compiled | RegexOptions.CultureInvariant);

    /// <summary>
    /// Longest-first so KZN is preferred over a trailing two-letter match.
    /// </summary>
    private static readonly string[] ProvinceSuffixes =
    [
        "KZN",
        "EC", "FS", "GP", "LP", "MP", "NC", "NW", "WC", "WP"
    ];

    public static string Normalize(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return string.Empty;
        }

        return value
            .Trim()
            .Replace(" ", string.Empty, StringComparison.Ordinal)
            .Replace("-", string.Empty, StringComparison.Ordinal)
            .ToUpperInvariant();
    }

    public static bool IsValid(string? value)
    {
        var normalized = Normalize(value);

        if (normalized.Length is < MinLength or > MaxLength)
        {
            return false;
        }

        if (!normalized.All(static c => c is >= 'A' and <= 'Z' or >= '0' and <= '9'))
        {
            return false;
        }

        if (LegacyPattern.IsMatch(normalized))
        {
            return true;
        }

        if (!TrySplitProvince(normalized, out var body))
        {
            return false;
        }

        if (body.Length is < 1 or > 8 || !body.Any(char.IsLetter))
        {
            return false;
        }

        // Newer provincial: 2–8 alphanumeric with at least one digit.
        if (body.Length is >= 2 and <= 8 && body.Any(char.IsDigit))
        {
            return true;
        }

        // Personalised: 1–7 alphanumeric with at least one letter (digit optional).
        return body.Length <= 7;
    }

    public static void ValidateOrThrow(string? value)
    {
        if (!IsValid(value))
        {
            throw new DomainException(InvalidMessage);
        }
    }

    private static bool TrySplitProvince(string normalized, out string body)
    {
        foreach (var suffix in ProvinceSuffixes)
        {
            if (normalized.Length > suffix.Length &&
                normalized.EndsWith(suffix, StringComparison.Ordinal))
            {
                body = normalized[..^suffix.Length];
                return true;
            }
        }

        body = string.Empty;
        return false;
    }
}

File: frontend/src/features/vehicles/registrationNumber.ts
const PROVINCE_SUFFIXES = ['KZN', 'EC', 'FS', 'GP', 'LP', 'MP', 'NC', 'NW', 'WC', 'WP'] as const;

export const REGISTRATION_MAX_LENGTH = 12;
export const REGISTRATION_INVALID_MESSAGE = 'Registration number must be a valid South African plate.';

/** Uppercase and strip spaces/hyphens — mirrors backend RegistrationNumber.Normalize. */
export function normalizeRegistrationNumber(value: string): string {
  return value.trim().replace(/[\s-]/g, '').toUpperCase();
}

function splitProvince(normalized: string): string | null {
  for (const suffix of PROVINCE_SUFFIXES) {
    if (normalized.length > suffix.length && normalized.endsWith(suffix)) {
      return normalized.slice(0, -suffix.length);
    }
  }
  return null;
}

/**
 * Broad SA plate check: legacy provincial, newer provincial + suffix, personalised + suffix.
 * Mirrors Bruno.Domain.ValueObjects.RegistrationNumber.IsValid.
 */
export function isValidSouthAfricanRegistration(value: string): boolean {
  const normalized = normalizeRegistrationNumber(value);

  if (normalized.length < 3 || normalized.length > REGISTRATION_MAX_LENGTH) {
    return false;
  }

  if (!/^[A-Z0-9]+$/.test(normalized)) {
    return false;
  }

  if (/^[A-Z]{1,3}\d{1,6}$/.test(normalized)) {
    return true;
  }

  const body = splitProvince(normalized);
  if (!body || body.length < 1 || body.length > 8 || !/[A-Z]/.test(body)) {
    return false;
  }

  // Newer provincial: 2–8 alphanumeric with at least one digit.
  if (body.length >= 2 && body.length <= 8 && /\d/.test(body)) {
    return true;
  }

  // Personalised: 1–7 alphanumeric with at least one letter.
  return body.length <= 7;
}
